### PR TITLE
Service Bus Queues and Topics and child resources

### PR DESCRIPTION
**Pull request by bjorkstromm**
WIP PR for #1789

Open issues for the Bicep provisioner explaned below.

Since Bicep lacks support for nested loops I can't do something like this
```bicep
param queues array = []
param topics object = {}

resource serviceBusNamespace 'Microsoft.ServiceBus/namespaces@2022-10-01-preview' = {
  ...

  resource queue 'queues@2022-10-01-preview' = [for name in queues:{
    name: name
  }]

  resource topic 'topics@2022-10-01-preview' = [for topic in items(topics):{
    name: topic.key
    
    resource subscription 'subscriptions@2022-10-01-preview' = [for subscription in topic.value.subscriptions: {
      name: subscription.name
    }]
  }]
}
```

And instead I'd need a module for the topics (see https://stackoverflow.com/a/76105656).

To solve this, should we 
a) Make `AzureBicepServiceBusTopic` (and maybe Queue also) inherit from `AzureBicepResource` and create a separate Bicep files. Haven't looked into the Bicep provisioner, but hopefully I could have the topic resource to depend on the Service Bus namespace?
b) Add support for dependent modules in `BicepTemplateFile` and make `AzureBicepResource.GetBicepTemplateFile` emit all needed files to disk upon provisioning.
c) Try to get everything working in a single Bicep file. Might not even be doable, and it sure won't be pretty :smile:

//cc @davidfowl 
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2232)

### CODE REVIEW

**Comment by davidfowl** [comment]
I would leave the non-bicep version as is. We plan to change those to be all bicep based.

File: src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs [hunk 1]
```diff
@@ -29,6 +34,63 @@ public class AzureBicepServiceBusResource(string name) :
     public string? GetConnectionString() => ServiceBusEndpoint.Value;
 }
 
+/// <summary>
+/// Represents an Azure Service Bus Queue resource.
+/// </summary>
+public class AzureBicepServiceBusQueueResource(string name, AzureBicepServiceBusResource parent) :
+    Resource(name),
+    IResourceWithConnectionString,
```
Comment (bjorkstromm) [inline src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs:42]: Should this really implement `IResourceWithConnectionString`? The connection string is on the namespace level.

File: src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs [hunk 2]
```diff
@@ -29,6 +34,63 @@ public class AzureBicepServiceBusResource(string name) :
...
+        context.Writer.WriteString("parent", Parent.Name);
+    }
+}
+
+/// <summary>
+/// Represents an Azure Service Bus Queue resource.
+/// </summary>
+public class AzureBicepServiceBusTopicResource(string name, string[] subscriptions, AzureBicepServiceBusResource parent) :
+    Resource(name),
+    IResourceWithConnectionString,
```
Comment (bjorkstromm) [inline src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs:68]: Should this really implement `IResourceWithConnectionString`? The connection string is on the namespace level.

File: src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs [hunk 3]
```diff
@@ -39,20 +101,94 @@ public static class AzureBicepServiceBusExtensions
...
-        // TODO: Change topics and queues to child resources
 
         return builder.AddResource(resource)
                       .WithParameter("serviceBusNamespaceName", resource.CreateBicepResourceName())
                       .WithParameter(AzureBicepResource.KnownParameters.PrincipalId)
                       .WithParameter(AzureBicepResource.KnownParameters.PrincipalType)
-                      .WithParameter("topics", topicNames ?? [])
-                      .WithParameter("queues", queueNames ?? [])
+                      // TODO: Shouldn't these be lazily evaluated?
+                      .WithParameter("queues", resource.Queues)
```
Comment (bjorkstromm) [inline src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs:114]: We need to find a nice way to make these lazily evaluated.

File: src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs [hunk 4]
```diff
@@ -39,20 +101,94 @@ public static class AzureBicepServiceBusExtensions
...
+                      // TODO: Shouldn't these be lazily evaluated?
+                      .WithParameter("topics", new JsonArray(resource.Topics.Select(CreateTopicJsonObject).ToArray()))
```
Comment (bjorkstromm) [inline src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs:116]: We need to find a nice way to make these lazily evaluated.

File: src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs [hunk 5]
```diff
@@ -39,20 +101,94 @@ public static class AzureBicepServiceBusExtensions
...
+    }
+
+    private static IResourceBuilder<AzureBicepServiceBusQueueResource> AddQueue(this IDistributedApplicationBuilder builder, string name, AzureBicepServiceBusResource parent)
+    {
+        var resource = new AzureBicepServiceBusQueueResource(name, parent);
+
+        parent.Queues.Add(name);
+
+        // TODO: This should not be needed here. It should be lazily evaluated.
+        parent.Parameters["queues"] = parent.Queues;
```
Comment (bjorkstromm) [inline src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs:147]: Remove this once parameters are lazily evaluated.

File: src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs [hunk 6]
```diff
@@ -39,20 +101,94 @@ public static class AzureBicepServiceBusExtensions
...
+    }
+
+    private static IResourceBuilder<AzureBicepServiceBusTopicResource> AddTopic(this IDistributedApplicationBuilder builder, string name, string[] subscriptions, AzureBicepServiceBusResource parent)
+    {
+        var resource = new AzureBicepServiceBusTopicResource(name, subscriptions, parent);
+
+        parent.Topics.Add(name, subscriptions);
+
+        // TODO: This should not be needed here. It should be lazily evaluated.
+        parent.Parameters["topics"] = new JsonArray(parent.Topics.Select(CreateTopicJsonObject).ToArray());
```
Comment (bjorkstromm) [inline src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs:182]: Remove this once parameters are lazily evaluated.

File: src/Aspire.Hosting.Azure/Bicep/servicebus.bicep
Comment (bjorkstromm) [inline src/Aspire.Hosting.Azure/Bicep/servicebus.bicep:1]: Managed to get it working with single file. It didn't become too ugly 😅

File: src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs [hunk 7]
```diff
@@ -39,20 +101,94 @@ public static class AzureBicepServiceBusExtensions
...
+    /// Adds a Azure Service Bus Queue resource to the application model. This resource requires an <see cref="AzureServiceBusResource"/> to be added to the application model.
+    /// </summary>
+    /// <param name="serviceBusBuilder">The Azure Service Bus resource builder.</param>
+    /// <param name="name">The name of the queue.</param>
+    public static IResourceBuilder<AzureBicepServiceBusQueueResource> AddQueue(this IResourceBuilder<IResourceWithParent<AzureBicepServiceBusResource>> serviceBusBuilder, string name)
+    {
+        return serviceBusBuilder.ApplicationBuilder.AddQueue(name, serviceBusBuilder.Resource.Parent);
+    }
+
+    private static IResourceBuilder<AzureBicepServiceBusQueueResource> AddQueue(this IDistributedApplicationBuilder builder, string name, AzureBicepServiceBusResource parent)
```
Comment (bjorkstromm) [inline src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs:140]: Should we just return `IResourceBuilder<AzureBicepServiceBusResource>`

File: src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs [hunk 8]
```diff
@@ -39,20 +101,94 @@ public static class AzureBicepServiceBusExtensions
...
+    /// </summary>
+    /// <param name="serviceBusBuilder">The Azure Service Bus resource builder.</param>
+    /// <param name="name">The name of the topic.</param>
+    /// <param name="subscriptions">The name of the subscriptions.</param>
+    public static IResourceBuilder<AzureBicepServiceBusTopicResource> AddTopic(this IResourceBuilder<IResourceWithParent<AzureBicepServiceBusResource>> serviceBusBuilder, string name, string[] subscriptions)
+    {
+        return serviceBusBuilder.ApplicationBuilder.AddTopic(name, subscriptions, serviceBusBuilder.Resource.Parent);
+    }
+
+    private static IResourceBuilder<AzureBicepServiceBusTopicResource> AddTopic(this IDistributedApplicationBuilder builder, string name, string[] subscriptions, AzureBicepServiceBusResource parent)
```
Comment (bjorkstromm) [inline src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs:175]: Should we just return `IResourceBuilder<AzureBicepServiceBusResource>`

**Review by bjorkstromm — COMMENTED** [review]
@davidfowl I managed to solve the deployment in a single Bicep file with a little map and reduce. If it's ok, we'll leave the module support for another PR 😄

**Comment by davidfowl** [comment]
> @davidfowl I managed to solve the deployment in a single Bicep file with a little map and reduce. If it's ok, we'll leave the module support for another PR 😄

Or course.

File: src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs (hunk 1 above)
Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs:42]: Nope. I don't think this needs to be a resource at all.

File: src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs (hunk 2 above)
Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs:68]: I'm not sure this should be a resource either.

File: src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs (hunk 3 above)
Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs:114]: @mitchdenny just added a callback API

File: src/Aspire.Hosting.Azure/Bicep/servicebus.bicep [hunk 9]
```diff
@@ -12,6 +12,8 @@ param queues array = []
 param topics array = []
 
 var resourceToken = uniqueString(resourceGroup().id)
+var topicNames = map(topics, topic => topic.name)
+var subscriptions = reduce(map(topics, topic => map(topic.subscriptions, sub => { name: sub, topic: topic.name })), [], (acc, item) => concat(acc,item))
```
Comment (davidfowl) [inline src/Aspire.Hosting.Azure/Bicep/servicebus.bicep:16]: oh cool haskell

**Comment by davidfowl** [comment]
Also this is going in https://github.com/dotnet/aspire/pull/2320.

File: src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs (hunk 8 above)
Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Bicep/AzureBicepServiceBusResource.cs:175]: Yep

**Comment by davidfowl** [comment]
Hey @bjorkstromm do you think you can get it in this week? We're locking down for p4 and I'd love to have this change.

**Comment by bjorkstromm** [comment]
> Hey @bjorkstromm do you think you can get it in this week? We're locking down for p4 and I'd love to have this change.

Sure thing! I'll make it work.

**Comment by bjorkstromm** [comment]
@davidfowl ready for review. Don't think the test Helix failures on Linux is related to this change. Although I don't really know how to check 😅 
https://dev.azure.com/dnceng-public/public/_build/results?buildId=575796&view=logs&j=22fea640-1099-5f32-ec5d-316ad83f359a&t=dad169e2-35ab-5ac6-55b3-7eed7d423255

**Comment by davidfowl** [comment]
This change is really good! (as usual @bjorkstromm!)

**Review by davidfowl — APPROVED** [review]

**Comment by davidfowl** [comment]
Thanks!

**Pull request merged**